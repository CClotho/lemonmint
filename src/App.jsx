import {Link, Outlet} from 'react-router-dom'
import { useState, useEffect } from 'react'
import { getProducts } from './api/api';
import {useNavigate} from 'react-router-dom';
import { RiShirtFill } from "react-icons/ri";
import { GiPearlNecklace } from "react-icons/gi";
import { FaLemon } from "react-icons/fa";
import styles from './App.module.css';


function App() {
  const [products, setProducts] = useState([]);
  const [cart, setCart] = useState([]);
  const [quantity, setQuantity] = useState(null);
  const [onEdit, setOnEdit] = useState(null);
  const [totalPrice, setTotalPrice] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);

  const navigate = useNavigate();

  useEffect(() => {

   setError(null);
    const AllProducts = async() => {
      try {
        const allProducts = await getProducts();
        setProducts(allProducts)
        
      }catch(error) {
        setError(error)
      }finally{
        setIsLoading(false);
      }
    }

    AllProducts();
  }, [])  

  
 
 // console.log("App Products",products)


 
  const onAddProduct =(id) => {
    
    const addProduct = products.find(product => {
      if(product.id === id) {
        return {...product, quantity: 1};
      }
      
    })

    console.log("Add product", cart)
    if(addProduct) {

      setCart([...cart, addProduct]);
      setQuantity(1)
      computeTotalPrice();
     
    }
   

    return addProduct;

    
    
  }

 
  const computeTotalPrice= () => {
    let products = 0;
    
    const total = cart.map(product => {
    
    
    products += product.price
    
    return products
    })

    setTotalPrice(products)
    
  }


  const onHandleCheckout = () => {
    
    //computeTotalPrice(); add this on useEffect on checkout component
    
    navigate("/checkout")
  }

 
  const onRemoveProduct =(id) => {
    
    const updatedCart = cart.filter((product, index) => index !== id)
     console.log(updatedCart)
    if(updatedCart) {
      setCart(updatedCart);
      return true;
    }
    else {
      return false;
    }
    
    
  }

  const onChangeQuantity = (userQuantity) => {
    setQuantity(userQuantity);
  }

 
  const onUpdateCart =(id, quantity, setQuantity) => {
    // edit this part to look for the index of the product inside the cart and update it rather than product id because if 2 ids are in the cart it will share the same quantity
    const updatedCart = cart.map((product, index) => {

      
      let newPrice = product.price * quantity
      if(id === index) {
        
        return {...product, quantity: quantity, price: newPrice }
      } else {
        return product
      }
      
      
    })

    setCart(updatedCart);
    setOnEdit(null)
    setQuantity(null)
    
    
  }


  console.log(products)

  return (
    <>
      <div className={styles.container}>
        <div>
        <Link to="/" className={styles.links}> <h1> Welcome to Lemonmint Shop</h1></Link>
        </div>
          {/* Add hamburger icon here*/}
     
        <div className={styles.dropdown}>
          <FaLemon size="2.5rem" cursor="pointer" className={styles.hidden}/>

         <div className={styles.menu}>
          
          <div >
          <Link to="/clothes" className={styles.links}> <RiShirtFill size="2rem" cursor="pointer"/></Link>
            <Link to="/clothes" className={styles.links}> Clothes</Link>
          </div>

          
          <div>
          <Link to="/jewelry" className={styles.links}>  <GiPearlNecklace size="2rem" cursor="pointer"/></Link>
            <Link to="/jewelry" className={styles.links}> Jewelry</Link>
          </div>
         </div>

        </div>

        
        <div>
          <div  onClick={onHandleCheckout} className={styles.links}>   Cart: {cart.length} </div>
        </div>


    
      </div>

        
    <Outlet context={[
      onAddProduct,
      products,
      cart,
      onRemoveProduct,
      onUpdateCart,
      computeTotalPrice,
      totalPrice,
      quantity,
      onEdit,               
      setOnEdit,
      onChangeQuantity
    ]}
      />

   </>
  )
}

export default App
