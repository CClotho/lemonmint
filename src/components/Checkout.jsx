import { useEffect, useState } from "react";
import {useOutletContext} from "react-router-dom";


export const Cart = ()=> {
    
    // You have to use all of the shared value in the parent context can't only choose products, cart etc like it has to be in sequence else use ",, " to skip items
    const [,, cart, onRemoveProduct, onUpdateCart, computeTotalPrice, totalPrice, , onEdit, setOnEdit] = useOutletContext();
    const [quantity, setQuantity] = useState(null);
    console.log(cart)
    
    useEffect(() => {

        computeTotalPrice();


    },[cart])
    return(

        <div> 
        <h1> Checkout</h1>

       <div>
       {Array.isArray(cart) && cart?.length > 0 ? cart.map((product, index) => {
            return(
                <div key={index}>
               <ul>
                    <li><img src={product.image ? product.image : "-"} alt={product.title}/></li>
                    <li>{product.title}</li>
                    <li>{product.description}</li>
                    <li>{product.price}</li>
                    <button type="button" onClick={()=> onRemoveProduct(index)}> Remove Item</button>
                    <button type="button" onClick={()=> setOnEdit(index)}> Update Quantity</button>
                    {onEdit === index ? (
                        <>
                            <input type="number" value={quantity > 0 ? quantity : product.quantity} onChange={(e)=> setQuantity(e.target.value)}/>
                            <button type="button" onClick={() => onUpdateCart(index, quantity, setQuantity)}>Confirm</button>
                            <button type="button" onClick={() => setOnEdit(null)}>Cancel</button>
                        </>
                    ) :   <li>{product.quantity}</li> }

                </ul>      
            </div>
            )
        }):
         <div> You currently do not have items in your cart </div>
        
        }
       </div>

       <div>
        <h2> Checkout: {totalPrice}</h2>
       </div>
    </div>

    )
}