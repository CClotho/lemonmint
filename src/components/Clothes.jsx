import { useEffect } from "react";
import { useOutletContext } from "react-router-dom";
import { GiCutLemon } from "react-icons/gi";
import styles from '../App.module.css';

export const Clothes = () => {

    const [onAddProduct, products, cart, , , computeTotalPrice]= useOutletContext();
    
    const Clothes = products.filter(product => product.category === "men's clothing");

    useEffect(() => {

       // computeTotalPrice();
    },[])
   
     console.log(Clothes)
    return(
        <>

        <h1 className={styles.categoryTitle }> Welcome to Men's clothing</h1>
            <div className={styles.itemsContainer}> 
                

                {Clothes && Clothes?.length > 0 ? Clothes.map(product => {
                    return(
                        <div key={product.id}>
                    <ul className={styles.card}>
                            <li><img src={product.image} alt={product.title} className={styles.shirts}/></li>
                            <li>{product.title}</li>
                            <li>{product.description}</li>
                            <li>{product.price}</li>
                            <button type="button" onClick={()=> onAddProduct(product.id)} className={styles.addBtn}>  <GiCutLemon size="2rem"  className={styles.lemon}/>Add to cart</button>

                        </ul>      
                    </div>
                    )
                }):
                <div> No products available </div>
                
                }
            </div>
        </>
    )
}