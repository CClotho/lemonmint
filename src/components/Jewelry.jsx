import {useOutletContext} from 'react-router-dom';
import { useEffect } from 'react';
import { GiCutLemon } from "react-icons/gi";
import styles from '../App.module.css';

export const Jewelry = ()=> {

        const [onAddProduct, products, cart, , , computeTotalPrice] = useOutletContext();

        const Jewelry = products.filter(product => product.category === 'jewelery');   
        
        useEffect(() => {

            computeTotalPrice();
        },[])
        
    return(
        <>
            <h1 className={styles.categoryTitle }> Welcome to Jewelry </h1>
            <div className={styles.itemsContainer}> 
            

                {Jewelry?.length > 0 ? Jewelry.map(product => {
                    return(
                        <div key={product.id}>
                        
                        <ul className={styles.card}>
                            <li><img src={product.image} alt={product.title} className={styles.accessories}/></li>
                            <li>{product.title}</li>
                            <li>{product.description}</li>
                            <li>{product.price}</li>
                            <button type="button" onClick={()=> onAddProduct(product.id)} className={styles.addBtn}>  <GiCutLemon size="2rem"  className={styles.lemon}/>Add to cart</button>

                        </ul>             
                    </div>
                    )
                }):
                <div> No products available </div>
                
                }

            </div>
        </>
    )
}