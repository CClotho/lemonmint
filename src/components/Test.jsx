import { useEffect } from "react";
import {useOutletContext} from "react-router-dom";

export const Cart = ()=> {
    
    // You have to use all of the shared value in the parent context can't only choose products, cart etc like it has to be in sequence else use ",, " to skip items
    const [,, cart, onRemoveProduct, onUpdateCart, computeTotalPrice, totalPrice, quantity, onEdit, setOnEdit, onChangeQuantity] = useOutletContext();
   
    console.log(cart)
    
    useEffect(() => {

        computeTotalPrice();


    },[cart])
    return(

        <div> 
        <h1> Checkout</h1>

       <div>
       {Array.isArray(cart) && cart?.length > 0 ? cart.map((product, index) => {
            return(
                <div key={index}>
               <ul>
                    <li><img src={Array.isArray(product.images) ? product.images[0] : product.images} alt={product.title}/></li>
                    <li>{product.title}</li>
                    <li>{product.description}</li>
                    <li>{product.price}</li>
                    <button type="button" onClick={()=> onRemoveProduct(product.id)}> Remove Item</button>
                    <button type="button" onClick={()=> setOnEdit(product.id)}> Update Quantity</button>
                    {onEdit === product.id ? (
                        <>
                            <input type="number" value={quantity > 0 ? quantity : 1} onChange={(e)=> onChangeQuantity(e.target.value)}/>
                            <button type="button" onClick={() => onUpdateCart(product.id)}>Confirm</button>
                        </>
                    ) :   <li>{product.quantity}</li> }

                </ul>      
            </div>
            )
        }):
         <div> You currently do not have items in your cart </div>
        
        }
       </div>

       <div>
        <h2> Checkout: {totalPrice}</h2>
       </div>
    </div>

    )
}


const Component = (props) => { // Accessing the object property by dot notation

    return (
        <div>
            {props.name}  {/* Accessing the 'name' property using dot notation */}
        </div>
    )

}


const Component2 = ({name}) => { // accessing object's property by object destructuring
    return (
        <div>
            {name} {/* Accessing the 'name' property directly through object destructuring */}
        </div>
    )
}


export const DisplayComponents = () => {


    return(
        <>
            <Component name={"Kay"}/> {/* you can technically give any name for your object property that you will eventually pass something in your component*/}
            <Component2 name={"Ray"}/> {/* here though you use the object destructuring as your parameter so you can directly access the object property's directly by their property name*/}
        </>
    )
}


