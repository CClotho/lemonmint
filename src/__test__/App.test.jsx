import {vi, describe, it, expect} from 'vitest';
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Clothes } from '../components/Clothes';
import {Cart} from '../components/Checkout';
import { BrowserRouter } from 'react-router-dom';
import App from '../App';


vi.mock('../components/Checkout', () => ({
    __esModule: true,
    Cart: () => <div>Cart Page</div>
}))

vi.mock('../components/Clothes', () => ({
    __esModule: true,
    Clothes: () => <div>Clothes Page</div>
  }));

describe('../App', () => {
    it('It should render the homepage', () => {

     const {container} = render(
            <BrowserRouter>
            <App/>
            
            </BrowserRouter>);
    
            const welcomeMessage = screen.getByText('Welcome to Lemonmint Shop');
    
            // Assert that the element is in the document
            expect(welcomeMessage).toBeInTheDocument(); 
          
    })

    

    it('It should render the right page according to the link (route) that is clicked', async() => {
        const user = userEvent.setup(); 
        render(
        <BrowserRouter>
            <App/>   
        </BrowserRouter>);

        const clothesPage = screen.getByRole('link',{name: 'Clothes'});

        await user.click(clothesPage);  

        const {container} = render(<Clothes/>);

        expect(container).toMatchSnapshot();
       
    }) 

    it('It should navigate to cart when the Cart button is clicked', async() => {
        const user = userEvent.setup(); 
       
        

        render(<BrowserRouter>
            <App/>
        </BrowserRouter>)

        const CartButton = screen.getByText(/cart/i);

        await user.click(CartButton);

        const {container} = render(<Cart/>);

        expect(container).toMatchSnapshot();


    })
})
