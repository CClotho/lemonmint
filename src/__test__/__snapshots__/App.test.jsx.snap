// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`../App > It should navigate to cart when the Cart button is clicked 1`] = `
<div>
  <div>
    Cart Page
  </div>
</div>
`;

exports[`../App > It should render the right page according to the link (route) that is clicked 1`] = `
<div>
  <div>
    Clothes Page
  </div>
</div>
`;

exports[`App Component > It should navigate to cart when the Cart button is clicked 1`] = `
<div>
  <div>
    Cart Page
  </div>
</div>
`;

exports[`App Component > It should render the homepage 1`] = `<div />`;

exports[`App Component > It should render the right page according to the link (route) that is clicked 1`] = `
<div>
  <div>
    Clothes Page
  </div>
</div>
`;
